r/R/util.R

```diff
@@ -110,3 +110,36 @@ handle_embedded_nul_error <- function(e) {
   }
   stop(e)
 }
+
+#' Take an object of length 1 and repeat it.
+#' 
+#' @param object Object of length 1 to be repeated - vector, `Scalar`, `Array`, or `ChunkedArray`
+#' @param n Number of repetitions
+#' 
+#' @return `Array` of length `n`
+#' 
+#' @keywords internal
+repeat_value_as_array <- function(object, n) {
+  if (inherits(object, "ChunkedArray")) {
+    return(MakeArrayFromScalar(Scalar$create(object$chunks[[1]]), n))
+  }
+  return(MakeArrayFromScalar(Scalar$create(object), n))
+}
+
+#' Recycle scalar values in a list of arrays
+#' 
+#' @param arrays List of arrays
+#' @return List of arrays with any vector/Scalar/Array/ChunkedArray values of length 1 recycled 
+#' @keywords internal
+recycle_scalars <- function(arrays){
+  # Get lengths of items in arrays
+  is_df <- map_lgl(arrays, ~inherits(.x, "data.frame"))
+  arr_lens <- lengths(arrays)
+  arr_lens[is_df] <- map_int(arrays[is_df], nrow)
+  
+  if (length(arrays) > 1 && any(arr_lens == 1) && !all(arr_lens==1)) {
+    max_array_len <- max(arr_lens)
+    arrays[arr_lens == 1 & !is_df] <- lapply(arrays[arr_lens == 1 & !is_df], repeat_value_as_array, max_array_len)
```

Comment:
Why not data frames? they turn into struct arrays:

```
> Scalar$create(tibble::tibble(a=1))$as_array()
StructArray
<struct<a: int32>>
-- is_valid: all not null
-- child 0 type: int32
  [
    1
  ]
``

Reply:
RecordBatch__from_arrays fails with a `StructArray` as input due to `count_fields` (from recordbatch.cpp) checking if the object inherits from a data.frame - I've tried to implement something in `18be171`, but it's wrong, and I'm struggling to figure out how to do it correctly.

Reply:
@romainfrancois can you take a look please?

Reply:
I've now ended up removing this extra code to try to handle mixed length tibble/data.frames as it seems like a lot of effort for something that doesn't seem to be a particularly common use case.  I've added in an error message for if the input tables are tibbles of different lengths.